internals/get-package-json/src/index.ts

```diff
@@ -0,0 +1,41 @@
+import fs from 'fs';
+import path from 'path';
+
+function getPackageJSONPath(dir: string) {
+  return path.join(dir, 'package.json');
+}
+
+export function getPackageJSON() {
+  const _prepareStackTrace = Error.prepareStackTrace;
+
+  Error.prepareStackTrace = (_, stack) => stack;
+
+  // The `stack` reference above is of type `Array<NodeJS.CallSite>`
+  const stack = new Error().stack as unknown as Array<NodeJS.CallSite>;
+
+  Error.prepareStackTrace = _prepareStackTrace;
+
+  // TypeScript rule noUncheckedIndexedAccess makes `stack[1]` potentially undefined;
+  // however, we can safely assert it is not undefined as the callstack will always
+  // have at least 2 entries, the first being this function, and the second wherever
+  // its being called from.
+  const callsite = stack[1]!;
+
+  // it is guarranteed that the only way for `getFileName` to return `undefined` is if
+  // this function is called using `eval` thus it is safe to assert `filePath` is defined
+  // at this point.
+  const filePath = (callsite.getFileName() ||
+    callsite.getEvalOrigin()) as string; // get the file name of where this function was called
+
+  let rootDir = path.dirname(filePath);
+
+  let packageJSONPath = getPackageJSONPath(rootDir);
+  while (!fs.existsSync(packageJSONPath)) {
+    rootDir = path.join(rootDir, '..');
+    packageJSONPath = getPackageJSONPath(rootDir);
+  }
+
+  const packageJSON = JSON.parse(fs.readFileSync(packageJSONPath, 'utf-8'));
+
+  return packageJSON;
```

Comment:
IMO having this function do the casting is dangerous. Unless we add JSON validation, the best return type for this function is `any` or `unknown`. Because authors may _think_ we are validating when we really aren't and it could introduce hard to find bugs. Leaving the casting up to the user of the function is best because then it is clear in their code where that type comes from (and thus if there are issues they can quickly see "oh i'm type casting when I should be type validating").

Reply:
Also, replacing `pkg.ts` with this function call everywhere is going to be a perf hit, as every code path that uses it would have to re-execute this entire function every time they need a value from package.json

It would be cool to add some kind of auto caching/memoization so that this function could be reused in a code base, but not have to find and read the package.json file every time.

Reply:
Added caching 😄

Reply:
> IMO having this function do the casting is dangerous.

It is a weaker typing, that's true.

> replacing pkg.ts with this function call everywhere is going to be a perf hit

Oh, you're doing a one-time execution in `pkg.ts` and returning the result. I didn't notice that. Yeah, it's worth keeping in that form.

> Added caching 😄

Sounds good!